sdk/core/dma/dma_compartment.cc

```diff
@@ -0,0 +1,196 @@
+         *  so that it can wait for timeout via passing this value  
+         */
+        if (previousInterruptCounter == currentInterruptCounter) 
+        {
+            return -(currentInterruptCounter + 1);
+        } 
+
+        /*
+         *  Potentially, redundant futex wait and context switch
+         *  todo: do we wanna remove it and let the thread pass?
+         */
+        uint32_t *dmaIsLaunchedPtr = &dmaIsLaunched;
+        futex_wait(dmaIsLaunchedPtr, dmaIsLaunched);
+        
+    } 
+
+    /**
+     *  Designate that the dmaIsLaunched,
+     *  if the thread accessed the dma first
+     */
+    dmaIsLaunched = 1;
+
+    /**
+     *  After acquiring a ownership over lock,
+	 *  now it is the time to launch dma.
+     * 
+     *  Here, we claim the memory with default malloc capability,
+     *  as declaring another heap capability is an extra entry 
+     *  that leaves the default capability let unused.
+     *
+     *  Unique pointers are used to avoid explicit heap_free() calls.
+     *  when this pointer is not used anymore, it is automatically deleted
+     */
+    
+    auto claim = [](void  *ptr) -> std::unique_ptr<char>
+    {
+        if (heap_claim(MALLOC_CAPABILITY, ptr) == 0)
+        {
+            return {nullptr};
+        }
+        
+        return std::unique_ptr<char> { static_cast<char*>(ptr)};
+    };
+    
+    claimedSource = claim(sourceAddress);
+    claimedDestination = claim(targetAddress);
+   
+    if (!claimedSource || !claimedDestination)
+    {
+        return -EINVAL;
+    }
+
+    /**
+     *  return if sufficient permissions are not present 
+     *  and if not long enough 
+     */
+
+    if (!check_pointer<PermissionSet{Permission::Load, Permission::Global}>(sourceAddress, lengthInBytes) ||
+           !check_pointer<PermissionSet{Permission::Store, Permission::Global}>(targetAddress, lengthInBytes) )
+    {
+        return -EINVAL;
+    }
+
+    platformDma.write_conf_and_start(sourceAddress, targetAddress, lengthInBytes, 
+                                        sourceStrides, targetStrides, byteSwapAmount);         
+
+    /**
+     *  return here, if all operations 
+     *  were successful.
+     *  Save the current interrupt counter to the previous one
+     *  to differentiate betweent the start and end afterwards 
+     */
+    
+    previousInterruptCounter = currentInterruptCounter;
+    return currentInterruptCounter;;
+
+}
+
+void reset_and_clear_dma(uint32_t interruptNumber)
+{   
+    /**
+     *  Resetting the claim pointers 
+     *  and cleaning up the dma registers.
+     *
+     *  Claim registers are meant to be 
+     *  cleared by every DMA operation,
+     *  that is why we are explicitely resetting 
+     *  them at this exit function.
+     */
+
+    /**
+     *  Flip the dmaIsRunnning value to show that launch is finished,
+     *  in case finishing thread wakes from the interrupt first 
+     */
+    dmaIsLaunched = 0;
```

Comment:
This needs to hold the lock or a concurrent call here will corrupt state with a launch.